File: source/State/AbstractState.java
package source.State;

import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.math.Vector2;

/**
 * Convenience implementation of {@link State}. Derive from this and only override what you need.
 * However, {@link State#update(float)}}, {@link State#render(Batch)} and {@link State#dispose()} must be overridden.
 *
 * @author Xico
 */

abstract public class AbstractState implements State {

    protected AbstractState() {
    }

    @Override
    abstract public void update(float delta);

    @Override
    abstract public void render(Batch batch);

    @Override
    abstract public void dispose();

    @Override
    final public Vector2 unproject(float screenX, float screenY) {
        return unproject(new Vector2(screenX, screenY));
    }

    @Override
    final public Vector2 unproject(Vector2 screenCoordinates) {
        return StateManager.getInstance().getViewport().unproject(screenCoordinates);
    }

    @Override
    public void resize(int width, int height) {
    }

    @Override
    public void pause() {
    }

    @Override
    public void resume() {
    }

    @Override
    public void hide() {
    }

    @Override
    public boolean keyDown(int keycode) {
        return false;
    }

    @Override
    public boolean keyUp(int keycode) {
        return false;
    }

    @Override
    public boolean keyTyped(char c) {
        return false;
    }

    @Override
    public boolean touchDown(int x, int y, int pointer, int button) {
        return false;
    }

    @Override
    public boolean touchUp(int x, int y, int pointer, int button) {
        return false;
    }

    @Override
    public boolean touchDragged(int x, int y, int pointer) {
        return false;
    }

    @Override
    public boolean mouseMoved(int screenX, int screenY) {
        return false;
    }

    @Override
    public boolean scrolled(int amount) {
        return false;
    }

}

File: kot/state/AbstractState.kt
package kot.state

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.Vector2

abstract class AbstractState : State {

 /* State methods that absolutely must be overriden. */

  override abstract fun update(delta: Float)
  override abstract fun render(batch: Batch)
  override abstract fun dispose()

  /* State methods that may or may not be overriden. */

  override fun resize(width: Int, height: Int) {}
  override fun hide() {}
  override fun pause() {}
  override fun resume() {}

  /* State input conversion methods - can't be overriden. */

  final override fun unproject(screenCoordinates: Vector2): Vector2 {
    // TODO
    return Vector2()
  }

  final override fun unproject(screenX: Float, screenY: Float): Vector2 {
    // TODO
    return Vector2()
  }

  /* InputProcessor methods that may or may not be overriden. */

  override fun keyDown(keycode: Int): Boolean = false
  override fun keyUp(keycode: Int): Boolean = false
  override fun keyTyped(character: Char): Boolean = false
  override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean = false
  override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean = false
  override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean = false
  override fun mouseMoved(screenX: Int, screenY: Int): Boolean = false
  override fun scrolled(amount: Int): Boolean = false

}